File: dotnet/kubernetes/auto-instrumentation/aspnetcore/Controllers/DemoController.cs
using Microsoft.AspNetCore.Mvc;
using System.Diagnostics;

namespace DotnetApp.Controllers;

[ApiController]
[Route("/")]
public class DemoController : ControllerBase
{
    private readonly ILogger<DemoController> _logger;

    public DemoController(ILogger<DemoController> logger)
    {
        _logger = logger;
    }

    [HttpGet]
    public IActionResult GetRoot()
    {
        _logger.LogInformation("Root endpoint accessed");

        // Get current activity (span) created by auto-instrumentation
        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("endpoint", "/");
            activity.SetTag("method", "GET");
        }

        return Ok(new
        {
            message = "Hello from Sematext OpenTelemetry Example!",
            service = Environment.GetEnvironmentVariable("OTEL_SERVICE_NAME") ?? "dotnet-aspnetcore-k8s-auto",
            instrumentation = "auto"
        });
    }

    [HttpGet("api/users/{id}")]
    public async Task<IActionResult> GetUser(string id)
    {
        _logger.LogInformation("Fetching user with ID: {UserId}", id);

        // Get current activity and add custom tags
        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("user.id", id);
            activity.SetTag("operation", "get_user");
        }

        // Simulate database call
        await Task.Delay(100);

        _logger.LogInformation("Successfully retrieved user {UserId}", id);

        return Ok(new
        {
            id = id,
            name = $"User {id}",
            email = $"user{id}@example.com"
        });
    }

    [HttpGet("api/slow")]
    public async Task<IActionResult> GetSlow()
    {
        _logger.LogInformation("Slow endpoint called - simulating delay");

        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("operation.type", "slow");
        }

        // Simulate slow operation
        await Task.Delay(2000);

        _logger.LogInformation("Slow operation completed");

        return Ok(new { message = "This took 2 seconds!" });
    }

    [HttpGet("api/error")]
    public IActionResult GetError()
    {
        _logger.LogError("Error endpoint called - throwing intentional error");

        var activity = Activity.Current;
        if (activity != null)
        {
            activity.SetTag("error.intentional", true);
        }

        throw new Exception("Intentional error for testing traces");
    }
}

File: php/kubernetes/auto-instrumentation/laravel/app/Http/Controllers/DemoController.php
<?php

namespace App\Http\Controllers;

use Illuminate\Http\JsonResponse;
use OpenTelemetry\API\Trace\Span;
use OpenTelemetry\API\Trace\StatusCode;

class DemoController
{
    public function health(): string
    {
        return 'healthy';
    }

    public function ready(): string
    {
        return 'ready';
    }

    public function root(): JsonResponse
    {
        // Get current span (created automatically by auto-instrumentation)
        $span = Span::getCurrent();
        $span->setAttribute('custom.attribute', 'root-endpoint');

        return response()->json([
            'message' => 'Hello from Sematext OpenTelemetry Example!',
            'service' => env('OTEL_SERVICE_NAME', 'php-laravel-k8s-auto'),
            'instrumentation' => 'auto',
        ]);
    }

    public function getUser(string $userId): JsonResponse
    {
        // Get current span and add custom attributes
        $span = Span::getCurrent();
        $span->setAttribute('user.id', $userId);
        $span->setAttribute('operation', 'get_user');

        // Simulate database call
        usleep(100000); // 100ms

        return response()->json([
            'id' => $userId,
            'name' => "User {$userId}",
            'email' => "user{$userId}@example.com",
        ]);
    }

    public function slow(): JsonResponse
    {
        $span = Span::getCurrent();
        $span->setAttribute('operation.type', 'slow');

        // Simulate slow operation
        sleep(2);

        return response()->json(['message' => 'This took 2 seconds!']);
    }

    public function error(): JsonResponse
    {
        $span = Span::getCurrent();
        $span->setAttribute('error.intentional', true);

        try {
            throw new \RuntimeException('Intentional error for testing traces');
        } catch (\Throwable $e) {
            $span->recordException($e);
            $span->setStatus(StatusCode::STATUS_ERROR, $e->getMessage());

            return response()->json(['error' => $e->getMessage()], 500);
        }
    }
}
